src/test/java/lambda/part3/exercise/Exercise2.java

```diff
@@ -0,0 +1,156 @@
+            return new MapHelper<>(source);
+        }
+
+        public List<T> getMapped() {
+            return new ArrayList<>(source);
+        }
+
+        /**
+         * Создает объект для маппинга, передавая ему новый список, построенный на основе исходного.
+         * Для добавления в новый список каждый элемент преобразовывается с использованием заданной функции.
+         * ([T], (T -> R)) -> [R]
+         * @param mapping Функция преобразования элементов.
+         */
+        public <R> MapHelper<R> map(Function<T, R> mapping) {
+           List<R> result = new ArrayList<>();
+           source.forEach(T-> {
+               result.add(mapping.apply(T));
+           });
+            return new MapHelper<>(result);
+        }
+
+        /**
+         * Создает объект для маппинга, передавая ему новый список, построенный на основе исходного.
+         * Для добавления в новый список каждый элемент преобразовывается в список с использованием заданной функции.
+         * ([T], (T -> [R])) -> [R]
+         * @param flatMapping Функция преобразования элементов.
+         */
+        public <R> MapHelper<R> flatMap(Function<T, List<R>> flatMapping) {
+            List<List<R>> listOfListsOfR = new ArrayList<>();
+            source.forEach(T-> {
+                listOfListsOfR.add(flatMapping.apply(T));
+            });
+            List<R> result = new ArrayList<>();
+            listOfListsOfR.forEach(S-> {
+                S.forEach(L-> {
+                    result.add((R)L);
+                });
+            });
+            return new MapHelper<>(result);
+        }
+    }
+
+    @Test
+    public void mapEmployeesToLengthOfTheirFullNamesUsingMapHelper() {
+        List<Employee> employees = getEmployees();
+
+        List<Integer> lengths = MapHelper.from(employees).map(Employee::getPerson).map(Person::getFullName).map(String::length).getMapped();
+        // TODO                 MapHelper.from(employees)
+        // TODO                          .map(Employee -> Person)
+        // TODO                          .map(Person -> String(full name))
+        // TODO                          .map(String -> Integer(length of string))
+        // TODO                          .getMapped();
+        assertEquals(Arrays.asList(14, 19, 14, 15, 14, 16), lengths);
+    }
+
+    @Test
+    public void mapEmployeesToCodesOfLetterTheirPositionsUsingMapHelper() {
+        List<Employee> employees = getEmployees();
+
+        List<Integer> codes = MapHelper.from(employees).flatMap(Employee::getJobHistory).map(JobHistoryEntry::getPosition).
+                flatMap(String->{char[] arrayOfChars = String.toCharArray();
+                    List<Character> listOfCharacters = new ArrayList<>();
+                    for (char aChar : arrayOfChars) {
+                        listOfCharacters.add(aChar);
+                    }
+                    return listOfCharacters;}).map(Character->(int)Character).getMapped();
```

Comment:
It's better to extract multilined lambdas into local variable or separate method.
If no, you could format code clearer, something like:
```Java
yourStreamWithActions
.flatMap(string -> {
    char[] arrayOfChars = string.toCharArray();
    List<Character> listOfCharacters = new ArrayList<>();
    for (char aChar : arrayOfChars) {
        listOfCharacters.add(aChar);
    }
    return listOfCharacters;
})
.map(character -> (int) character)
.getMapped();
```

Reply:
Ok. I extracted this code to separated method. To be honest, I couldn't find solution to convert string to list of characters directly. I had to create the array of chars as interim variable.